crates/uv-requirements/src/workspace.rs

```diff
@@ -0,0 +1,705 @@
+/// │               └── bar.py
+/// ├── pyproject.toml
+/// ├── Readme.md
+/// └── uv.lock
+/// ```
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(serde::Serialize))]
+pub struct ProjectWorkspace {
+    /// The path to the project root.
+    project_root: PathBuf,
+    /// The name of the package.
+    project_name: PackageName,
+    /// The workspace the project is part of.
+    workspace: Workspace,
+}
+
+impl ProjectWorkspace {
+    /// Find the current project and workspace.
+    pub fn discover(path: impl AsRef<Path>) -> Result<Self, WorkspaceError> {
+        let Some(project_root) = path
+            .as_ref()
+            .ancestors()
+            .find(|path| path.join("pyproject.toml").is_file())
+        else {
+            return Err(WorkspaceError::MissingPyprojectToml);
+        };
+
+        debug!(
+            "Found project root: `{}`",
+            project_root.simplified_display()
+        );
+
+        Self::from_project_root(project_root)
+    }
+
+    /// The directory containing the closest `pyproject.toml`, defining the current project.
+    pub fn project_root(&self) -> &Path {
+        &self.project_root
+    }
+
+    /// The name of the current project.
+    pub fn project_name(&self) -> &PackageName {
+        &self.project_name
+    }
+
+    /// The workspace definition.
+    pub fn workspace(&self) -> &Workspace {
+        &self.workspace
+    }
+
+    /// Return the requirements for the project.
+    pub fn requirements(&self) -> Vec<RequirementsSource> {
+        vec![
+            RequirementsSource::from_requirements_file(self.project_root.join("pyproject.toml")),
+            RequirementsSource::from_source_tree(self.project_root.clone()),
+        ]
+    }
+
+    fn from_project_root(path: &Path) -> Result<Self, WorkspaceError> {
+        let pyproject_path = path.join("pyproject.toml");
+
+        // Read the `pyproject.toml`.
+        let contents = fs_err::read_to_string(&pyproject_path)?;
+        let pyproject_toml: PyProjectToml = toml::from_str(&contents)
+            .map_err(|err| WorkspaceError::Toml(pyproject_path.clone(), err))?;
+
+        // Extract the `[project]` metadata.
+        let Some(project) = pyproject_toml.project.clone() else {
+            return Err(WorkspaceError::MissingProject(pyproject_path));
+        };
+
+        Self::from_project(path.to_path_buf(), &pyproject_toml, project.name)
+    }
+
+    /// Find the workspace for a project.
+    fn from_project(
+        project_path: PathBuf,
+        project: &PyProjectToml,
+        project_name: PackageName,
+    ) -> Result<Self, WorkspaceError> {
+        let mut workspace = project
+            .tool
+            .as_ref()
+            .and_then(|tool| tool.uv.as_ref())
+            .and_then(|uv| uv.workspace.as_ref())
+            .map(|workspace| (project_path.clone(), workspace.clone(), project.clone()));
+
+        if workspace.is_none() {
+            workspace = find_workspace(&project_path)?;
```

Comment:
Should this be an `or_else`?

Reply:
Does this mean we're in a _member_ of a workspace?

Reply:
yeah, added some docs in the other branch